storage/remote/client.go

```diff
@@ -198,7 +209,28 @@ func (c *Client) Store(ctx context.Context, req []byte) error {
 		err = errors.Errorf("server returned HTTP status %s: %s", httpResp.Status, line)
 	}
 	if httpResp.StatusCode/100 == 5 {
-		return RecoverableError{err}
+		return RecoverableError{err, retryAfter{duration: noRetryAfter}}
+	}
+	if httpResp.StatusCode == 429 {
+		recErr := RecoverableError{err, retryAfter{duration: noRetryAfter}}
+		if after := httpResp.Header.Get("Retry-After"); after != "" {
+			duration, err := time.Parse(time.RFC1123, after)
+			if err == nil {
+				recErr.retryAfter = retryAfter{isDateTime: true, after: duration}
+			} else {
+				// The duration can be simply in seconds.
+				secs, errAtoi := strconv.Atoi(after)
+				if errAtoi != nil {
+					return errors.New("Retry-After must be either in seconds or in date-time")
+				}
+				parsedDuration, err := time.ParseDuration(fmt.Sprintf("%ds", secs))
+				if err != nil {
+					return errors.Wrap(err, "unable to parse duration")
+				}
+				recErr.retryAfter.duration = model.Duration(parsedDuration)
+			}
+		}
```

Comment:
Let's break this code out into a separate function that checks for the retry-after header and returns a duration (which would be 0 in the case of a missing header or error).